### PR TITLE
Fix shadowing/same-name warnings for later declarations

### PR DESCRIPTION
**Pull request by a3d4**
```
function e() {}
contract C {
    function f() pure public { uint e; uint g; uint h; }
    function g() pure public {}
}
function h() {}
```

The old implementation warns about shadowing `e()` only. This PR adds warnings for `g()` and `h()`.

### CODE REVIEW

File: libsolidity/analysis/NameAndTypeResolver.h [hunk 1]
```diff
@@ -44,6 +44,8 @@ class ErrorReporter;
 namespace solidity::frontend
 {
 
+using DeferredDeclarations = std::vector<std::tuple<std::string, langutil::SourceLocation, std::shared_ptr<DeclarationContainer const>>>;
```
Comment (a3d4) [inline libsolidity/analysis/NameAndTypeResolver.h]: I did not look at lifetimes yet to check if regular pointers would be sufficient.

File: libsolidity/analysis/NameAndTypeResolver.h [hunk 2]
```diff
@@ -126,6 +131,9 @@ class NameAndTypeResolver: private boost::noncopyable
 	DeclarationContainer* m_currentScope = nullptr;
 	langutil::ErrorReporter& m_errorReporter;
 	GlobalContext& m_globalContext;
+	// For declarations we save name, location and declaration container.
+	// Then, after all declarations are known, we process the list warn about declarations with the same name.
```
Comment (chriseth) [inline libsolidity/analysis/NameAndTypeResolver.h]:
```suggestion
	// Then, after all declarations are known, we process the list and warn about declarations with the same name.
```

Comment (chriseth) [reply libsolidity/analysis/NameAndTypeResolver.h]: Please use triple slashes.

File: libsolidity/analysis/NameAndTypeResolver.cpp [hunk 3]
```diff
@@ -219,6 +219,49 @@ void NameAndTypeResolver::warnVariablesNamedLikeInstructions()
 	}
 }
 
+void NameAndTypeResolver::warnHomonymDeclarations()
+{
+	for (auto const& decl: m_deferredDeclarations)
+	{
+		string name = get<0>(decl);
```
Comment (chriseth) [inline libsolidity/analysis/NameAndTypeResolver.cpp]: This sounds like you actually want this to be a struct :)

File: libsolidity/analysis/NameAndTypeResolver.h (hunk 1 above)
Comment (chriseth) [reply libsolidity/analysis/NameAndTypeResolver.h]: the lifetimes are bound to NameAndTypeResolver's lifetime which is destroyed at the end of the compilation process.

**Comment by chriseth** [comment]
Having to pass `m_deferredDeclarations` all over the place does not look very nice. Furthermore, the container we register the declaration at is already passed to the functions.

Instead of `m_deferredDeclarations`, we could store pointers to child containers in the container and then walk the structure of containers and maybe even make the homonym search a function of the container?

**Comment by a3d4** [comment]
> Having to pass `m_deferredDeclarations` all over the place does not look very nice.

I completely agree, it looks redundant. The issues are that in `DeclarationRegistrationHelper::registerDeclaration`
- `_errorLocation` might differ from `_declaration.location()`,
- `_warnOnShadow` flag needs to bee preserved.

So it seems to me that (deferred) declarations need to be decorated somehow with that information. I could add parameters `_errorLocation` and `_warnOnShadow` to `DeclarationContainer::registerDeclaration`, store them somehow, and at the end walk the container tree. (The order in which we report the shadowing warnings does not  matter, right?)

**Comment by chriseth** [comment]
Yeah, maybe add another structure to DeclarationContainer that contains the error locations and "warn on shadow" flags and so on instead of maintaining another parameter.

Actually - is `warn on shadow` set to false anywhere else except for the point where we re-register `this` and `super`? Maybe that can be simplified?

File: libsolidity/analysis/DeclarationContainer.h [hunk 4]
```diff
@@ -38,17 +40,26 @@ namespace solidity::frontend
...
 	explicit DeclarationContainer(
 		ASTNode const* _enclosingNode = nullptr,
-		DeclarationContainer const* _enclosingContainer = nullptr
+		DeclarationContainer* _enclosingContainer = nullptr
 	):
-		m_enclosingNode(_enclosingNode), m_enclosingContainer(_enclosingContainer) {}
+		m_enclosingNode(_enclosingNode), m_enclosingContainer(_enclosingContainer)
+	{
+		if (_enclosingContainer)
+			_enclosingContainer->m_innerContainers.emplace_back(this);
```
Comment (a3d4) [inline libsolidity/analysis/DeclarationContainer.h:50]: This is fragile. We have to make sure that no `DeclarationContainer`s get destructed earlier.

For instance, [`DeclarationRegistrationHelper::enterNewSubScope`](https://github.com/ethereum/solidity/pull/9877/files#diff-653a1e23aa75715b13289b03e2f20947R612) was rewritten to avoid temporary containers.

File: libsolidity/analysis/DeclarationContainer.cpp [hunk 5]
```diff
@@ -164,3 +184,19 @@ vector<ASTString> DeclarationContainer::similarNames(ASTString const& _name) con
 
 	return similar;
 }
+
+void DeclarationContainer::populateHomonyms(back_insert_iterator<Homonyms> _it) const
```
Comment (chriseth) [inline libsolidity/analysis/DeclarationContainer.cpp:188]: This could also just be
```suggestion
Homonyms DeclarationContainer::homonyms() const
```

File: libsolidity/analysis/DeclarationContainer.cpp [hunk 6]
```diff
@@ -164,3 +184,19 @@ vector<ASTString> DeclarationContainer::similarNames(ASTString const& _name) con
...
+{
+	for (DeclarationContainer const* innerContainer: m_innerContainers)
+		innerContainer->populateHomonyms(_it);
+
+	for (auto [name, location]: m_homonymCandidates)
+	{
+		Declaration const* homonymDeclaration = nullptr;
+		for (auto const* declaration: m_enclosingContainer->resolveName(name, true, true))
+			homonymDeclaration = declaration;
```
Comment (chriseth) [inline libsolidity/analysis/DeclarationContainer.cpp]: Any reason for not adding a `break` after the first?

**Review by chriseth — CHANGES_REQUESTED** [review]
Please add a changelog entry.

File: libsolidity/analysis/DeclarationContainer.cpp (hunk 5 above)
Comment (a3d4) [reply libsolidity/analysis/DeclarationContainer.cpp:188]: In an intermediate implementation, I had something like public `Homonyms homonyms()` wrapping private `void populateHomonyms(....)`. That makes the interface cleaner, but requires an extra wrapper, _which is used just once_. So after some thoughts I decided against it. (No strong opinion here though, can go back to it.)

The root cause is that I do not see an easy way to avoid recursive populating of `Homonyms` list.

File: libsolidity/analysis/DeclarationContainer.cpp (hunk 6 above)
Comment (a3d4) [reply libsolidity/analysis/DeclarationContainer.cpp]: Historical reasons, it is what we did before. But I wondered too :).

Actually, I thought it might make more sense to process all declarations, not just first or last. I did not want to address it within this PR, but since you brought it, please look at 27ffffb1952d087295b31e5e44e12a33f4ca1dd1.

Comment (chriseth) [reply libsolidity/analysis/DeclarationContainer.cpp]: I think it's better to only report one shadowed declaration.

File: test/libsolidity/syntaxTests/multiSource/free_function_resolution_override_virtual.sol [hunk 7]
```diff
@@ -16,4 +16,5 @@ contract D is C {
 // ----
 // Warning 2519: (s1.sol:65-134): This declaration shadows an existing declaration.
 // Warning 2519: (s2.sol:85-155): This declaration shadows an existing declaration.
+// Warning 2519: (s2.sol:85-155): This declaration shadows an existing declaration.
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/multiSource/free_function_resolution_override_virtual.sol]: I don't think this is really helpful.

File: Changelog.md [hunk 8]
```diff
@@ -45,8 +46,8 @@ Bugfixes:
  * Code generator: Fix internal error on stripping dynamic types from return parameters on EVM versions without ``RETURNDATACOPY``.
  * Type Checker: Add missing check against nested dynamic arrays in ABI encoding functions when ABIEncoderV2 is disabled.
  * Type Checker: Correct the error message for invalid named parameter in a call to refer to the right argument.
- * Type Checker: Correct the warning for homonymous, but not shadowing declarations.
  * Type Checker: Disallow ``virtual`` for modifiers in libraries.
+ * Name Resolver: Correct the warning for homonymous, but not shadowing declarations.
```
Comment (chriseth) [inline Changelog.md:70]: Is this a correction or here by accident?

File: libsolidity/analysis/DeclarationContainer.cpp (hunk 5 above)
Comment (chriseth) [reply libsolidity/analysis/DeclarationContainer.cpp:188]: I don't understand why you would need a wrapper, but I also don't understand why recursion would be an issue here. Why doesn't the following work:
```suggestion
Homonyms DeclarationContainer::homonyms() const
{
	Homonyms ret;
	for (DeclarationContainer const* innerContainer: m_innerContainers)
		ret += innerContainer->homonyms();

	for (auto [name, location]: m_homonymCandidates)
		for (auto const* declaration: m_enclosingContainer->resolveName(name, true, true))
			ret.emplace_back(location, declaration);
	return ret;
}
```

Comment (a3d4) [reply libsolidity/analysis/DeclarationContainer.cpp:188]: > I also don't understand why recursion would be an issue here.

Sorry, I did not mention complexity. Let me rephrase:

> The root cause is that I do not see an easy way to avoid recursive populating of Homonyms list
> **which would give `O(N)` complexity**.

Even if quadratic time is not a practical issue here, it does not feel quite right :).

File: libsolidity/analysis/DeclarationContainer.cpp (hunk 6 above)
Comment (a3d4) [reply libsolidity/analysis/DeclarationContainer.cpp]: Why? It makes code slightly shorter and avoids the decision which name collision to report.

File: test/libsolidity/syntaxTests/multiSource/free_function_resolution_override_virtual.sol (hunk 7 above)
Comment (a3d4) [reply test/libsolidity/syntaxTests/multiSource/free_function_resolution_override_virtual.sol]: I guess it might be OK, because **solc** points at different shadowed declarations with different secondary locations. It is just that secondary locations are ignored by **isoltest**.

Comment (chriseth) [reply test/libsolidity/syntaxTests/multiSource/free_function_resolution_override_virtual.sol]: Shouldn't we add all shadowed declarations to the list of secondaries instead?

File: libsolidity/analysis/DeclarationContainer.cpp (hunk 5 above)
Comment (chriseth) [reply libsolidity/analysis/DeclarationContainer.cpp:188]: I would say that readability always trumps efficiency unless it is proven that the runtime is actually an issue, but we can also keep it like that.

File: Changelog.md (hunk 8 above)
Comment (a3d4) [reply Changelog.md:70]: No. Previous time I tagged similar fix as "Type Checker", while "Name Resolver" might be more accurate. I just noticed it now, while adding a new change log entry.

File: libsolidity/analysis/DeclarationContainer.cpp (hunk 5 above)
Comment (a3d4) [reply libsolidity/analysis/DeclarationContainer.cpp:188]: Sure, we should be aware of preliminary optimization. But in this concrete case, the code with quadratic complexity might disturb me more, making to think if it was really intended.

File: test/libsolidity/syntaxTests/multiSource/free_function_resolution_override_virtual.sol (hunk 7 above)
Comment (a3d4) [reply test/libsolidity/syntaxTests/multiSource/free_function_resolution_override_virtual.sol]: Yes. I did not realize that `SecondarySourceLocation` is a multi-location thing :).

I made a separate [commit](d9c49bd209e65a8cf01f4a9e81d82bf57c045f64)  to simplify review.

**Comment by chriseth** [comment]
Needs rebase.

**Review by chriseth — DISMISSED** [review]

**Comment by a3d4** [comment]
Rebased and added a couple of tests.

**Review by chriseth — APPROVED** [review]
